[UI] Fill the node information for node list table

ChengYanJin opened the pull request:
**Component**: ui, nodes

**Context**: Fill the node list table

**Summary**:
- Add the Control Plane IP and Workload Plane IP under the Node Name
- Get the Roles by the labels `node-role.kubernetes.io/<role>`
- Be able to retrieve the alerts from alertmanager.
- But haven't implemented the filter to compute the health
- the color of the status depends on the condition and status
          // green for status.conditions['Ready'] == True and all other conditions are false
          // yellow for status.conditions['Ready'] == True and some other conditions are true
          // red for status.conditions['Ready'] == False
          // grey when there is no status.conditions
- Update the unit test in nodes

**Acceptance criteria**: 
![image](https://user-images.githubusercontent.com/18453133/92754611-4ca35e00-f38b-11ea-92e9-1cf5c1b16cb7.png)


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2772 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/ducks/nodes: Fetch the IPs and Interface of the Nodes through SALT API

Action FETCH_NODES_IPS_INTERFACES is faster than SALT_AUTHENTICATION_SUCCESS, hence we get 401 Unauthorized here
Manually authenticate salt API before get the IPs and Interfaces for Nodes

Refs: #2772

ChengYanJin committed:
ui/services/salt: Fetch CP WP IPs and Interfaces from Salt API

Refs: #2772

ChengYanJin committed:
ui/utils: Add getNodeListData utility function for Node list table

Refs: #2772

ChengYanJin committed:
ui/services/alertmanger: Initialize the alertmanager service

We should retrieve the alerts from alertmanager by endpoint `/api/v2/alerts`

Refs: #2772

ChengYanJin committed:
ui/ducks/nodes: Add conditions field for nodes

The conditions contains the all the conditions to display in the node table

Refs: #2772

ChengYanJin committed:
ui/ducks/saga: Add the alertsSaga to rootSaga

Refs: #2772

ChengYanJin committed:
ui/ducks: Initialize the alertmanager api when set up the config

Refs: #2772

ChengYanJin committed:
ui/nodelist: Fill the node list table

Todo: Filter the alerts based on the Node and compute the Health

Refs: #2772

ChengYanJin committed:
ui/nodes: Add fetch ips interface and fetch alerts action

Refs: #2772

ChengYanJin committed:
ui/reducer: Combine the alert reducer

Refs: #2772

ChengYanJin committed:
ui/ducks/alerts: Add the alerts ducks

Refs: #2772

ChengYanJin committed:
ui/salt: Add the url_alertmanager in the config.json

Refs: #2772

ChengYanJin committed:
ui/nodes: Add the node list for nodelist table

Refs: #2772

ChengYanJin committed:
ui/nodes: Get the Roles of the Node from the kubect labels

the Roles of the Node should be the ones that are stored in the labels `node-role.kubernetes.io/<role-name>`

Refs: #2772

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2772-fill-the-node-information-for-node-list-table` and the
destination branch `development/2.6`.

Please resolve the conflict on **the feature branch** (`improvement/2772-fill-the-node-information-for-node-list-table`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2772-fill-the-node-information-for-node-list-table
 $ git merge origin/development/2.6
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2772-fill-the-node-information-for-node-list-table
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2772-fill-the-node-information-for-node-list-table` and the
destination branch `development/2.6`.

Please resolve the conflict on **the feature branch** (`improvement/2772-fill-the-node-information-for-node-list-table`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2772-fill-the-node-information-for-node-list-table
 $ git merge origin/development/2.6
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2772-fill-the-node-information-for-node-list-table
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

MonPote commented on ui/src/components/NodeListTable.js:
```diff
@@ -270,34 +303,72 @@ function Table({ columns, data, rowClicked, theme }) {
+          // yellow for status.conditions['Ready'] == True and some other conditions are true
+          // red for status.conditions['Ready'] == False
+          // grey when there is no status.conditions
+          const otherConditions = conditions?.filter(
```
```javascript
const otherConditions = conditions?.filter((cond) => cond !== 'ready') ?? [];
```

MonPote commented on ui/src/components/NodeListTable.js:
```diff
@@ -270,34 +303,72 @@ function Table({ columns, data, rowClicked, theme }) {
+                {intl.translate('not_ready')}
+              </StatusText>
+            );
+          } else if (!otherConditions) {
```
If you do the preview change then it's should be `} else`

gdemonet replied:
In any case, it makes sense to always show this "unknown" status as a fallback (especially given how the `status` prop is computed).

MonPote commented on ui/src/services/NodeUtils.js:
```diff
@@ -0,0 +1,31 @@
+        status: { status: node?.status, conditions: node?.conditions },
+        roles: node?.roles,
+      };
+    });
```
```javascript
    return nodes?.map((node) => {
      return {
        // IPs of Control Plane and Workload Plane are in the same Cell with Name
        name: {
          name: node?.name,
          control_plane_ip:
            (nodesIPsInterfaces[node.name] &&
              nodesIPsInterfaces[node.name][METALK8S_CONTROL_PLANE_IP]) ??
            undefined,
          workload_plane_ip:
            (nodesIPsInterfaces[node.name] &&
              nodesIPsInterfaces[node.name][METALK8S_WORKLOAD_PLANE_IP]) ??
            undefined,
        },
        status: { status: node?.status, conditions: node?.conditions },
        roles: node?.roles,
      };
    }) ?? [];
```

ChengYanJin committed:
ui/tests: Fix the uit test of config which was broken because of adding the alertmanager service

Refs: #2772

ChengYanJin committed:
ui/nodelist: Add empty array as default value for getNodelist

Refs: #2772

ChengYanJin committed:
ui/layout: Add Salt API authentication checking before load the pages components in private route

Otherwise we may get 401 Unauthorized when calling the Salt API

Refs: #2772

ChengYanJin committed:
ui/nodes/tests: Update the unit test for the Roles of node

Refs: #2772

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -439,6 +454,21 @@ export function* stopRefreshNodes() {
+export function* fetchNodesIPsInterface() {
+  // Action FETCH_NODES_IPS_INTERFACES is faster than SALT_AUTHENTICATION_SUCCESS, hence we get 401 Unauthorized here
+  // Manually authenticate salt API before get the IPs and Interfaces for Nodes
+  yield call(authenticateSaltApi);
```
Not sure that's the right way of waiting for some other action to complete (we always authenticate Salt API on login, so we shouldn't need to re-authenticate). Instead, your goal should be to model the behaviour of this specific action depending on the various cases:

- If we're authenticated to Salt API (`yield select(state => state.login.salt) !== null`), continue
- If we're not, wait for the authentication success/failure action

To manage this "waiting" behaviour, you can use some more advanced effect creators:

```javascript
const { success, failure, timeout } = yield race({
  success: take(SALT_AUTHENTICATION_SUCCESS),
  failure: take(SALT_AUTHENTICATION_FAILED),
  timeout: delay(3000),
});
```

See https://redux-saga.js.org/docs/advanced/RacingEffects.html for more info.

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -145,6 +149,7 @@ const defaultState = {
   list: [],
   isRefreshing: false,
   isLoading: false,
+  nodesIPsInterfaces: {},
```
This key lives under `state.app.nodes` already, so maybe we could find a shorter name... Maybe simply `IPs`, or `IPsInfo`?

And one other thing, we don't want to keep unneeded info in the application state, so let's reduce the interface as an IP attribute. In the end, I think this value should look like this (and it could be built in the reducer from the Salt API response):

```javascript
{ IPsInfo: {
  bootstrap: {
    control_plane: [
      { ip: '10.0.1.42', interface: 'eth1'}
    ],
    workload_plane: [
      { ip: '10.100.0.2', interface: 'eth3'},
      { ip: '10.105.1.3', interface: 'eth4'}
    ],
  },
  'master-1': {...},
} }
```

gdemonet commented on ui/src/services/NodeUtils.js:
```diff
@@ -0,0 +1,31 @@
+          workload_plane_ip:
+            (nodesIPsInterfaces[node.name] &&
+              nodesIPsInterfaces[node.name][METALK8S_WORKLOAD_PLANE_IP]) ||
+            undefined,
```
minor: Since you're checking if there is existing info about some node twice, would be better to store `nodesIPsInterfaces[node.name]` in a variable.

Also, small reminder that the "nullish coallescing operator" also works for bracket notation (cf. [the docs](https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining)).

And a small nitpick: let's not use `snake_case` in some places while `camelCase` is the "standard" in this codebase. Consistency is important!

Wrapping it all together:

```javascript
const IPsInfo = nodesIPsInterfaces?.[node.name];
return {
  name: {
    name: node.name,
    controlPlaneIP: IPsInfo?.[METALK8S_CONTROL_PLANE_IP],
    workloadPlaneIP: IPsInfo?.[METALK8S_WORKLOAD_PLANE_IP],
  },
  // ...
}
```

gdemonet commented on ui/src/services/alertmanager/api.js:
```diff
@@ -0,0 +1,11 @@
+}
+
+export function getAlerts() {
+  return alertmanagerApiClient.get('/api/v2/alerts');
```
Oh, small reminder that this `/v2` API is not stable yet. It's relatively OK to use it, because we will know when we update AlertManager, but maybe we should add some "stability" tests to check the API schema we rely on doesn't change without us noticing...

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -252,6 +252,17 @@ export function* fetchNodes() {
+          );
+          if (activeCondition && activeCondition.types) {
+            conditions.push(activeCondition.type);
+          }
```
That's only giving one condition, if any, in the `conditions` array... I guess it should be:

```suggestion
          const conditions = node.status.conditions.filter(
            (cond) => cond.status === 'True'
          ).map((cond) => cond.type);
```

Or, using a single `reduce` call (which allows to also check if `type` is a valid attribute):

```javascript
const conditions = node.status.conditions.reduce(
  (acc, cond) => {
    if (cond.status === 'True' && cond?.type)
      acc.push(cond.type);
    return acc;
  },
  []
);
```

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -252,6 +252,17 @@ export function* fetchNodes() {
               (conditon) => conditon.type === 'Ready',
             );
+
+          // Store the name of conditions which the status are True in the array
+          // given the avaiable conditions (Ready, DiskPressure, MemoryPressure, PIDPressure, Network Unavailable, Unschedulable)
```
Since we have all the information we need from the `Ready` condition, why do we bother with keeping it? If we look at how it's consumed in the nodes list, we're filtering it anyway (to get the `otherConditions`) :stuck_out_tongue_winking_eye:

ChengYanJin replied:
What you mean is pass the `condition` object directly to the `<NodeListTable>` component instead of creating conditions array with ['Ready'...]?

True! We did the filter in NodeListTable anyway. 
But actually there is a lot of information in the condition field, such as lastHeartbeatTime, lastTransitionTime, reason, message... those data is not useful for the <NodeListTable> 

So maybe it's better to create an array with the name of the active conditions and pass it to <NodeListTable>?

What do you think?

gdemonet replied:
Oh no that's not what I'm saying. I'm just suggesting that we could drop the `Ready` condition from this `conditions` array, because we don't look at it (since we have what we need in the `status` prop already) :slightly_smiling_face:

gdemonet commented on ui/src/containers/NodePage.js:
```diff
@@ -32,7 +41,10 @@ const NodePage = (props) => {
       </BreadcrumbContainer>
-      <NodePageContent nodes={nodes}></NodePageContent>
+      <NodePageContent
+        nodes={nodes}
+        nodeListData={nodeListData}
```
This is highly disturbing for a new-comer... We have `nodes`, which is supposedly a list of Node objects, and `nodeListData`, which is what? Some enriched list of Nodes with their associated alerts and IPs? Why not get only one of them?

ChengYanJin replied:
Right! It is very confusing. 
So maybe I should rename `nodeListData` to `nodeTableData` which is used by the NodeListTable component. 

The two object `nodeTableData` and `nodes` have some differences and we need both in the RightsSidePanel.

gdemonet replied:
What difference? Can't we make sure we have all we need in a single array?

ChengYanJin replied:
See for me, the `nodeTableData` should only include the data needed for the table, not more than that.
And the `nodes` is the object return by k8s with more information. 

Here actually we don't need to fetch the `nodes`. Since it's only needed in the health tabs. We should fetch it in the lower child component `<NodeRSP>`

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -271,21 +271,26 @@ export function* fetchNodes() {
+          let nodeRolesLabels = [];
           const roleTaintMatched = roleTaintMap.find((item) => {
-            const nodeRoles = Object.keys(node.metadata.labels).filter((role) =>
+            nodeRolesLabels = Object.keys(node.metadata.labels).filter((role) =>
               role.includes(ROLE_PREFIX),
             );
```
Instead of filling `nodeRolesLabels` on each pass of `find` (which is highly inefficient), just define it when creating the `nodeRolesLabels` variable (which should be a `const` anyway).

Also, can you make sure we don't forget to remove `roleTaintMatched`, `roleTaintMap` and the corresponding role props (`bootstrap`, `infra`, `control_plane` and `workload_plane`) once this page replaces the old one?

ChengYanJin replied:
I will create a ticket, once we are done with the New Node Page, we should change the current temporary URL to the `/volumes` and remove the unused props as you mentioned before `roleTaintMatched` and `roleTaintMatched`...

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -311,7 +316,7 @@ export function* fetchNodes() {
             bootstrap: roleTaintMatched && roleTaintMatched.bootstrap,
             infra: roleTaintMatched && roleTaintMatched.infra,
-            roles: rolesLabel.join(' / '),
+            roles: nodeRoles.join(' / '),
```
I'd rather make this an array, and let consumers display it as they wish. But since this prop is also consumed by the old page, I guess you can save this as debt.

gdemonet commented on ui/src/components/NodeListTable.js:
```diff
@@ -342,17 +342,17 @@ const NodeListTable = (props) => {
-            (cond) => cond !== 'ready',
+          const notReadyConditions = conditions?.filter(
+            (cond) => cond !== 'ready' ?? [],
           );
```
Not like this, or you'll get an array of arrays :stuck_out_tongue: 

```suggestion
          const notReadyConditions = conditions?.filter(
            (cond) => cond !== 'ready',
          )  ?? [];
```

gdemonet commented on ui/src/containers/PrivateRoute.js:
```diff
@@ -3,16 +3,22 @@ import { useSelector } from 'react-redux';
     return null;
-  } else {
+  } else if (isSaltAPIAuthenticated) {
+    // To make sure that Salt API is authenticated, before load the components
+    // Otherwise we may get 401 Unauthorized
```
That's a band-aid which is "covering" the issue but not fixing it. Actually, sagas which require Salt API shouldn't be called before we are authenticated. But components could very well be shown before that. And since most of our data doesn't come from Salt anyway, it makes a lot of sense to show it before we are authenticated to SaltAPI.

ChengYanJin replied:
You're absolutely right! There is nothing stopping us load the component. 
We should use `yield race` to wait for Salt Authenticated success action. 
It's great to discover this effect creator. 

Thank you!

gdemonet reviewed (changes requested):
When reworking sagas and the Redux state, let's try to tidy things up as much as possible, otherwise debt will start piling up and it will soon become a mess for anyone not used to it :)

Otherwise, most of the behaviour change looks good to me :+1:

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

ChengYanJin committed:
ui/ducks/nodes: Remove the unneeded info for nodesIPsInterfaces

- Rename it to IPsInfo
- Reduce the interface as an IP attribute

Refs: #2772

ChengYanJin committed:
ui/ducks/nodes: Wait until get the Salt api connected action before calling the Salt API

Salt API is not authenticated yet should not prevent us to load the components.

Refs: #2772

ChengYanJin committed:
ui/nodes: Fix the errors of the node condition

Refs: #2772

ChengYanJin committed:
ui/nodes: Rename the nodelistdata to nodetabledata

We don't need to fetch the nodes in NodePage level, since it's only used by Health tab.
We should fetch nodes in <NodeRSP>

Refs: #2772

ChengYanJin committed:
ui/nodes: Remove the Ready from the conditions

Refs: #2772

gdemonet commented on ui/src/services/NodeUtils.js:
```diff
@@ -0,0 +1,68 @@
+      ip: IPsInterfacesObject[METALK8S_CONTROL_PLANE_IP],
+      interface: Object.keys(IPsInterfacesObject[IP_INTERFACES]).find(
+        (en) =>
+          IPsInterfacesObject[IP_INTERFACES][en][0] ===
+          IPsInterfacesObject[METALK8S_CONTROL_PLANE_IP],
```
This kind of check assumes the IP we're looking for is returned first in `ip_interfaces`, which isn't robust... 

```suggestion
          IPsInterfacesObject[IP_INTERFACES][en].includes(
            IPsInterfacesObject[METALK8S_CONTROL_PLANE_IP]
          ),
```

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -439,6 +474,39 @@ export function* stopRefreshNodes() {
+  if (isSaltAPIAuthenticated) {
+    result = yield call(ApiSalt.getNodesIPsInterfaces);
+  } else {
+    // eslint-disable-next-line no-unused-vars
+    const { success, failure, timeout } = yield race({
```
It's sad that we are missing proper error-handling for this :confused: I'll open a debt ticket to track this.

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -246,21 +277,25 @@ export function* fetchNodes() {
           }
 
+          // the Roles of the Node should be the ones that are stored in the labels `node-role.kubernetes.io/<role-name>`
+          let nodeRolesLabels = [];
           const roleTaintMatched = roleTaintMap.find((item) => {
-            const nodeRoles = Object.keys(node.metadata.labels).filter((role) =>
+            nodeRolesLabels = Object.keys(node.metadata.labels).filter((role) =>
               role.includes(ROLE_PREFIX),
             );
```
Quoting my previous comment:

> Instead of filling `nodeRolesLabels` on each pass of `find` (which is highly inefficient), just define it when creating the `nodeRolesLabels` variable (which should be a `const` anyway).

That would look like:

```javascript
// the Roles of the Node should be the ones that are stored in the labels `node-role.kubernetes.io/<role-name>`
const nodeRolesLabels = Object.keys(node.metadata.labels).filter(
  (label) => label.startsWith(ROLE_PREFIX),
);
const roleTaintMatched = roleTaintMap.find((item) => {
  return (
    // what you already have
  );
};
```

Note that I used `startsWith` instead of `includes`, which is more accurate (`includes` could match a `custom.node-role.kubernetes.io/toto` label, which we don't want)

ChengYanJin replied:
Remove the `roleTaintMatched` and control_plane, workload_plane, bootstrap and infra from the returned Node object.
Since we can get the Roles from the Labels, I don't think we need those props anymore.

gdemonet reviewed (changes requested):
Last few changes and we're good to go!

ChengYanJin committed:
ui/ducks/nodes: Remove the unneeded props nodeRolesLabels and rolesLabel

Since we have labels to retrieve the Roles of Node, remove the control_plane,
workload_plane, bootstrap and infra from Node object

Refs: #2772

gdemonet commented on ui/src/ducks/app/nodes.test.js:
```diff
@@ -267,10 +263,6 @@ describe('`fetchNodes` saga', () => {
 
     const expectedNode = {
       ...defaultNodeForState,
-      workload_plane: expected.wp,
-      control_plane: expected.cp,
-      bootstrap: expected.bootstrap,
-      infra: expected.infra,
```
Since they're not used, you can remove them from the `expected` object

gdemonet reviewed (approved):
Perfect! Thanks for taking care of the debt :smiley:

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

ChengYanJin commented:
/approve

ChengYanJin commented:
> Perfect! Thanks for taking care of the debt 😃

I will. I think when tackle this debt ticket may deserve a workshop/discussion.

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/b17094d73f5f7853c08e70c96d591e1203236215) did not succeed in branch improvement/2772-fill-the-node-information-for-node-list-table.




*The following options are set:* **approve**

gdemonet commented:
> > Perfect! Thanks for taking care of the debt smiley
> 
> I will. I think when tackle this debt ticket may deserve a workshop/discussion.

Oh no I meant in this PR, no rush for the other one :wink:

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye chengyanjin.
